### ISSUE TITLE
Office365Connector action card issue :On submitting information if appplication run slow, it show error message "There was a problem submitting your changes. Try again in a minute."

### ISSUE THREAD

**Issue by poonam0025** [Microsoft/BotBuilder#4338]
## Bot Info
* SDK Platform: 

## Reproduction Steps
Run application in debug mode.
1. Launc offcie365Connector Action Card.
2. Fill required data and Click Submit button
3. There should debug point before submit operation .Wait debugger there up to 2 minute

## Expected Behavior
On click of Submit  data should save or in case of error user can handle the error.
****
**## Actual Results**
When application run slow /in debug mode
------------------------------------------------
**A pop up error message is showing**
"There was a problem submitting your changes. Try again in a minute."
Even submitted data is getting save . 
I want either I can hide this pop up error OR submitted data do not save after having that issue.

**Comment by stevengum97** [comment]
Please provide more information such as the code you're using, the SDK version and the SDK Language.

**Comment by poonam0025** [comment]
`private static Attachment CreateSampleO365ConnectorCard()
        {
            #region Multichoice Card
            var multichoiceCard = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Multiple Choice",
                "Multiple Choice Card",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "CardsType",
                        true,
                        "Pick multiple options",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Hero Card", "Hero Card"),
                            new O365ConnectorCardMultichoiceInputChoice("Thumbnail Card", "Thumbnail Card"),
                            new O365ConnectorCardMultichoiceInputChoice("O365 Connector Card", "O365 Connector Card")
                        },
                        "expanded",
                        true),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "Teams",
                        true,
                        "Pick multiple options",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Bot", "Bot"),
                            new O365ConnectorCardMultichoiceInputChoice("Tab", "Tab"),
                            new O365ConnectorCardMultichoiceInputChoice("Connector", "Connector"),
                            new O365ConnectorCardMultichoiceInputChoice("Compose Extension", "Compose Extension")
                        },
                        "compact",
                        true),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "Apps",
                        false,
                        "Pick an App",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("VSTS", "VSTS"),
                            new O365ConnectorCardMultichoiceInputChoice("Wiki", "Wiki"),
                            new O365ConnectorCardMultichoiceInputChoice("Github", "Github")
                        },
                        "expanded",
                        false),
                    new O365ConnectorCardMultichoiceInput(
                        O365ConnectorCardMultichoiceInput.Type,
                        "OfficeProduct",
                        false,
                        "Pick an Office Product",
                        null,
                        new List<O365ConnectorCardMultichoiceInputChoice>
                        {
                            new O365ConnectorCardMultichoiceInputChoice("Outlook", "Outlook"),
                            new O365ConnectorCardMultichoiceInputChoice("MS Teams", "MS Teams"),
                            new O365ConnectorCardMultichoiceInputChoice("Skype", "Skype")
                        },
                        "compact",
                        false)
            },

            new List<O365ConnectorCardActionBase>
                  {
                   new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "multichoice",
                        @"{""CardsType"":""{{CardsType.value}}"", ""Teams"":""{{Teams.value}}"", ""Apps"":""{{Apps.value}}"", ""OfficeProduct"":""{{OfficeProduct.value}}""}")
                 });

            #endregion

            #region Input Card
            var inputCard = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Text Input",
                "Input Card",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-1",
                        false,
                        "multiline, no maxLength",
                        null,
                        true,
                        null),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-2",
                        false,
                        "single line, no maxLength",
                        null,
                        false,
                        null),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-3",
                        true,
                        "multiline, max len = 10, isRequired",
                        null,
                        true,
                        10),
                    new O365ConnectorCardTextInput(
                        O365ConnectorCardTextInput.Type,
                        "text-4",
                        true,
                        "single line, max len = 10, isRequired",
                        null,
                        false,
                        10)
                },
                new List<O365ConnectorCardActionBase>
                {
                    new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "inputText",
                        @"{""text1"":""{{text-1.value}}"", ""text2"":""{{text-2.value}}"", ""text3"":""{{text-3.value}}"", ""text4"":""{{text-4.value}}""}")
                });
            #endregion

            #region Date Card
            var dateCard = new O365ConnectorCardActionCard(
                O365ConnectorCardActionCard.Type,
                "Date Input",
                "Date Card",
                new List<O365ConnectorCardInputBase>
                {
                    new O365ConnectorCardDateInput(
                        O365ConnectorCardDateInput.Type,
                        "date-1",
                        true,
                        "date with time",
                        null,
                        true),
                    new O365ConnectorCardDateInput(
                        O365ConnectorCardDateInput.Type,
                        "date-2",
                        false,
                        "date only",
                        null,
                        false)
                },
                new List<O365ConnectorCardActionBase>
                {
                    new O365ConnectorCardHttpPOST(
                        O365ConnectorCardHttpPOST.Type,
                        "Send",
                        "dateInput",
                        @"{""date1"":""{{date-1.value}}"", ""date2"":""{{date-2.value}}""}")
                });
            #endregion

            var section = new O365ConnectorCardSection(
                "**section title**",
                "section text",
                "activity title",
                "activity subtitle",
                "activity text",
                "http://connectorsdemo.azurewebsites.net/images/MSC12_Oscar_002.jpg",
                true,
                new List<O365ConnectorCardFact>
                {
                    new O365ConnectorCardFact("Fact name 1", "Fact value 1"),
                    new O365ConnectorCardFact("Fact name 2", "Fact value 2"),
                },
                new List<O365ConnectorCardImage>
                {
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/MicrosoftSurface_024_Cafe_OH-06315_VS_R1c.jpg",
                        Title = "image 1"
                    },
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/WIN12_Scene_01.jpg",
                        Title = "image 2"
                    },
                    new O365ConnectorCardImage
                    {
                        Image = "http://connectorsdemo.azurewebsites.net/images/WIN12_Anthony_02.jpg",
                        Title = "image 3"
                    }
                });

            O365ConnectorCard card = new O365ConnectorCard()
            {
                Summary = "O365 card summary",
                ThemeColor = "#E67A9E",
                Title = "card title",
                Text = "card text",
                Sections = new List<O365ConnectorCardSection> { section },
                PotentialAction = new List<O365ConnectorCardActionBase>
                 {
                    multichoiceCard,
                    inputCard,
                    dateCard,
                    new O365ConnectorCardViewAction(
                        O365ConnectorCardViewAction.Type,
                        "View Action",
                        null,
                        new List<string>
                        {
                            "http://microsoft.com"
                        })
                   
                  }
            };

            return card.ToAttachment();
        }

-----------------------
Use above card in application . After implementing above code  .Put debug point in application. Run application using ngrok in teams..Perform above card Send button action . and wait at debug point up to 2 minute .An error message in team comes intermittently "There was a problem submitting your changes. Try again in a minute."

**Comment by poonam0025** [comment]
![error](https://user-images.githubusercontent.com/37584638/37697655-0d31ad1c-2d04-11e8-983d-6216458d3bfc.PNG)

**Comment by nodeomega** [comment]
I am attempting to replicate the issue.  Is the error you were experiencing still recurring?

Also, does this issue occur when not debugging using ngrok?

**Comment by poonam0025** [comment]
yes, error is still occurring . This occurs when application runs slow. So this error occur in randomly. You  can re pro this error  by running you code in debug mode.  
For running code in debug mode I am using ngrok.exe

**Comment by satheeshganji48** [comment]
How to call API using O365ConnectorCardHttpPOST ..Is there any card type can be used for  the below potential actions 
 
**"potentialAction": [
                {
                    "@type": "ActionCard",
                    "name": "Approve",
                    "inputs": [
                        {
                            "@type": "TextInput",
                            "id": "comment",
                            "isMultiline": true,
                            "title": "Reason (optional)"
                        }
                    ],
                    "actions": [
                        {
                            "@type": "HttpPOST",
                            "name": "Submit",
                            "target": "https://{Site Address}/api/Approv?ID={{requestid}}",
                            "body": "={{comment.value}}",
                            "headers": [
                                {
                                    "Content-Type": "application/x-www-form-urlencoded"
                                }
                            ]
                        }
                    ]
                }**

**Comment by nodeomega** [comment]
@satheeshganji48 I am unclear what you are asking.  Is your question related to the "Send" button timeout that @poonam0025 was referring to?  This sounds like a separate issue altogether.

**Comment by nguyendn** [comment]
@nodeomega  How do you extend the timeout?

**Comment by nguyendn** [comment]
@nodeomega it seems Async is a solution, it works for me

```javascript
var o365CardActionAsyncHandler = async function(event, query, callback) {};
connector.onO365ConnectorCardAction(async o365CardActionAsyncHandler);

```

**Comment by nodeomega** [comment]
@nguyendn is your issue related to the send button timing out as well, and secondly that looks like NodeJS, correct?

**Comment by nguyendn** [comment]
@nodeomega yes. it related to the send button timing and it is NodeJS.

**Comment by nodeomega** [comment]
OK. @poonam0025  and @nguyendn ,   It would appear that there is no way to adjust that 10 second timeout for Invoke actions (the send button).  It would appear that this behavior is by design.

What I am told is that there are a few ways to workaround it however:
1. Ensure that your bot finishes processing the operation before the 10 second mark
2. Respond with HTTP 202 immediately but handle processing of the action in the background.

Also, if using IIS, be advised that the operation should complete before the threads time out, as IIS will do that after an extended amount of time.

As there is little more that can be done in terms of the Bot Framework itself in Teams regarding this issue, I will be closing this.  Thank you all for your patience regarding this, and I know this wasn't exactly the answer we were hoping for.
